### PR TITLE
Refactor/cache

### PR DESCRIPTION
**Pull request by Al-Qassim**
make caching for movies and series

I'm opening the pr before fixing the tests, so you can review the work

### CODE REVIEW

File: domain/src/main/java/com/cairosquad/domain/repository/SearchRepository.kt [hunk 1]
```diff
@@ -1,14 +1,8 @@
...
 import com.cairosquad.entity.Artist
-import com.cairosquad.entity.Movie
-import com.cairosquad.entity.Series
 
 interface SearchRepository {
-    suspend fun getSeries(query: String,page:Int): List<Series>
-
-    suspend fun getMovies(query: String,page:Int): List<Movie>
-
     suspend fun getArtists(query: String,page:Int): List<Artist>
```
Comment (ElsayedMagdy122) [inline domain/src/main/java/com/cairosquad/domain/repository/SearchRepository.kt]: Why did you leave function getArtists here , or is there no dedicated repo for it?

File: domain/src/main/java/com/cairosquad/domain/repository/SeriesRepository.kt [hunk 2]
```diff
@@ -9,7 +9,7 @@ import com.cairosquad.entity.Season
 import com.cairosquad.entity.Series
 
 interface SeriesRepository {
-    suspend fun getSeries(seriesId: Long): Series
+    suspend fun getSeriesById(seriesId: Long): Series
```
Comment (ElsayedMagdy122) [inline domain/src/main/java/com/cairosquad/domain/repository/SeriesRepository.kt]: Since the function's name is getSeriesById , I think if you leave the parameter name id only, it will be readable and simpler, and the function will express it.

File: domain/src/main/java/com/cairosquad/domain/repository/SeriesRepository.kt [hunk 3]
```diff
@@ -33,12 +33,13 @@ interface SeriesRepository {
 
     suspend fun getFreeToWatchSeries(page: Int,categoryId : String?): List<Series>
 
-    suspend fun getSeriesByCategory(category: String, page: Int): List<Series>
+    suspend fun getSeriesByCategory(genreId: String, page: Int): List<Series>
```
Comment (ElsayedMagdy122) [inline domain/src/main/java/com/cairosquad/domain/repository/SeriesRepository.kt]: Why is the function name getSeriesByCategory and the parameter genreId?

File: domain/src/main/java/com/cairosquad/domain/usecase/movies/GetMovieDetailsUseCase.kt [hunk 4]
```diff
@@ -10,7 +10,7 @@ class GetMovieDetailsUseCase(
 ) {
 
     suspend fun getMovie(movieId: Long): Movie {
```
Comment (ElsayedMagdy122) [inline domain/src/main/java/com/cairosquad/domain/usecase/movies/GetMovieDetailsUseCase.kt]: same here make the function name getMovieById

File: domain/src/main/java/com/cairosquad/domain/usecase/search/SearchUseCase.kt [hunk 5]
```diff
@@ -1,21 +1,25 @@
...
 import com.cairosquad.entity.Artist
 import com.cairosquad.entity.Movie
 import com.cairosquad.entity.Series
 
 class SearchUseCase(
     private val searchRepository: SearchRepository,
+    private val moviesRepository: MoviesRepository,
+    private val seriesRepository: SeriesRepository
 ) {
     suspend fun getSeries(query: String, page: Int): List<Series> {
```
Comment (ElsayedMagdy122) [inline domain/src/main/java/com/cairosquad/domain/usecase/search/SearchUseCase.kt]: make it getSeriesByQuery

File: domain/src/main/java/com/cairosquad/domain/usecase/search/SearchUseCase.kt [hunk 6]
```diff
@@ -1,21 +1,25 @@
...
-        return searchRepository.getSeries(query,page).also {
+        return seriesRepository.getSeriesByQuery(query,page).also {
             searchRepository.addQuery(query)
         }
     }
 
     suspend fun getMovies(query: String, page: Int ): List<Movie> {
```
Comment (ElsayedMagdy122) [inline domain/src/main/java/com/cairosquad/domain/usecase/search/SearchUseCase.kt]: same

File: domain/src/main/java/com/cairosquad/domain/usecase/series/GetSeriesDetailsUseCase.kt [hunk 7]
```diff
@@ -12,7 +12,7 @@ class GetSeriesDetailsUseCase(
 ) {
 
     suspend fun getSeries(seriesId: Long): Series {
```
Comment (ElsayedMagdy122) [inline domain/src/main/java/com/cairosquad/domain/usecase/series/GetSeriesDetailsUseCase.kt]: same

File: domain/src/test/java/com/cairosquad/domain/usecase/movies/GetMoviesDetailsUseCaseTest.kt [hunk 8]
```diff
@@ -25,12 +25,12 @@ class GetMoviesDetailsUseCaseTest {
 
     @Test
     fun `getMovie SHOULD return movie from repository`() = runTest {
-        coEvery { moviesRepository.getMovie(123L) } returns movie
+        coEvery { moviesRepository.getMovieById(123L) } returns movie
```
Comment (ElsayedMagdy122) [inline domain/src/test/java/com/cairosquad/domain/usecase/movies/GetMoviesDetailsUseCaseTest.kt]: id is hardCoded , you can create variable and pass it or is it duplicated in other test cases you can put in private companion object

File: local/src/main/java/com/cairosquad/local/cache/movie/MoviesCacheDao.kt [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+import com.cairosquad.repository.movie.data_source.local.dto.MovieGenreCacheCrossRef
+import com.cairosquad.repository.movie.data_source.local.dto.MovieWithoutGenreCacheDto
+import com.cairosquad.repository.movie.data_source.local.dto.RequestMovieCacheCrossRef
+import com.cairosquad.repository.movie.data_source.local.dto.RequestWithMoviesCacheDto
+
+@Dao
+interface MoviesCacheDao {
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun insertRequestMovieCacheCrossRef(mappings: List<RequestMovieCacheCrossRef>)
```
Comment (ElsayedMagdy122) [inline local/src/main/java/com/cairosquad/local/cache/movie/MoviesCacheDao.kt]: The name is long and not readable

File: local/src/main/java/com/cairosquad/local/cache/movie/MoviesCacheDao.kt [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+             "OR " +
+                "Not request in (Select request from RequestCacheDto)")
+    suspend fun deleteUnwantedRequestMovieCacheCrossRef()
+
+    @Query("Delete from MovieGenreCacheCrossRef " +
+            "where " +
+                "Not movie_id in (Select movie_id from MovieWithoutGenreCacheDto) " +
+             "OR " +
+                "Not genre_id in (Select genre_id from MovieGenreCacheCrossRef)")
+    suspend fun deleteUnwantedMovieGenreCacheCrossRef()
```
Comment (ElsayedMagdy122) [inline local/src/main/java/com/cairosquad/local/cache/movie/MoviesCacheDao.kt]: why the name is more long , In the end, it does not express enough ?

File: local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+import com.cairosquad.repository.utils.sharedDto.local.ReviewCacheDto
+
+class MoviesLocalDataSourceImpl(
+    private val moviesCacheDao: MoviesCacheDao,
+    private val requestCachedDao: RequestCachedDao,
+    private val genreDao: GenreDao,
+    private val reviewDao: ReviewDao
+): MoviesLocalDataSource {
+    override suspend fun insertRequestWithMovies(requestWithMovies: RequestWithMoviesCacheDto) {
+        moviesCacheDao.insertMoviesWithoutGenre(requestWithMovies.movies.map { it.movieWithoutGenre })
```
Comment (ElsayedMagdy122) [inline local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt]: format every extention function in new line and you can seprate it in small functions private

File: local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        genreDao.insertMovieGenres(genres = requestWithMovies.movies.map { it.genres }.flatten().toSet().toList())
```
Comment (ElsayedMagdy122) [inline local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt]: same

File: local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        moviesCacheDao.insertRequestMovieCacheCrossRef(
+            RequestMovieCacheCrossRef.fromRequestAndMovieList(
+                request = requestWithMovies.request,
+                movies = requestWithMovies.movies
+            )
+        )
+    }
+
+    override suspend fun getMoviesByRequest(request: String): List<MovieCacheDto> {
+
```
Comment (ElsayedMagdy122) [inline local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt]: remove this line and format this file

File: local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        reviewDao.insertReviews(requestWithReviewsCacheDto.reviews)
+        reviewDao.insertRequestReviewCacheCrossRef(
+            RequestReviewCacheCrossRef.fromRequestAndReviewList(
+                request = requestWithReviewsCacheDto.request,
+                reviews = requestWithReviewsCacheDto.reviews
+            )
+        )
+    }
+
+    override suspend fun deleteExpiredCache(timestamp: Long) {
```
Comment (ElsayedMagdy122) [inline local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt:79]: seprate implementation to small fuction private

File: local/src/main/java/com/cairosquad/local/cache/series/SeriesCacheDao.kt [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+import com.cairosquad.repository.series.data_source.local.dto.SeriesWithoutGenreCacheDto
+
+@Dao
+interface SeriesCacheDao {
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun insertRequestSeriesCacheCrossRef(mappings: List<RequestSeriesCacheCrossRef>)
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun insertSeriesGenreCacheCrossRef(mappings: List<SeriesGenreCacheCrossRef>)
```
Comment (ElsayedMagdy122) [inline local/src/main/java/com/cairosquad/local/cache/series/SeriesCacheDao.kt]: why parameter naming mappings ? it is not clear

File: local/src/main/java/com/cairosquad/local/cache/series/SeriesLocalDataSourceImpl.kt [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+import com.cairosquad.repository.utils.sharedDto.local.ReviewCacheDto
+
+class SeriesLocalDataSourceImpl(
+    private val seriesCacheDao: SeriesCacheDao,
+    private val requestCachedDao: RequestCachedDao,
+    private val genreDao: GenreDao,
+    private val reviewDao: ReviewDao
+): SeriesLocalDataSource {
+    override suspend fun insertRequestWithSeries(requestWithSeries: RequestWithSeriesCacheDto) {
+        seriesCacheDao.insertSeriesWithoutGenre(requestWithSeries.series.map { it.seriesWithoutGenre })
```
Comment (ElsayedMagdy122) [inline local/src/main/java/com/cairosquad/local/cache/series/SeriesLocalDataSourceImpl.kt]: seprate to small functions

File: local/src/main/java/com/cairosquad/local/cache/series/SeriesLocalDataSourceImpl.kt [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        reviewDao.insertReviews(requestWithReviewsCacheDto.reviews)
+        reviewDao.insertRequestReviewCacheCrossRef(
+            RequestReviewCacheCrossRef.fromRequestAndReviewList(
+                request = requestWithReviewsCacheDto.request,
+                reviews = requestWithReviewsCacheDto.reviews
+            )
+        )
+    }
+
+    override suspend fun deleteExpiredCache(timestamp: Long) {
```
Comment (ElsayedMagdy122) [inline local/src/main/java/com/cairosquad/local/cache/series/SeriesLocalDataSourceImpl.kt:76]: same seprate it

File: remote/src/main/java/com/cairosquad/remote/movie/MoviesRemoteDataSourceImpl.kt [hunk 18]
```diff
@@ -99,4 +104,14 @@ class RemoteMovieDataSourceImpl(
         return safeCallApi { apiService.getAllMovies(page,categoryId,sortBy) }
             .results?.filterNotNull().orEmpty()
     }
+
+    override suspend fun getPersonalizedMovies(page: Int): List<MovieRemoteDto> {
+        return safeCallApi { apiService.getPersonalizedMovies(page) }
```
Comment (ElsayedMagdy122) [inline remote/src/main/java/com/cairosquad/remote/movie/MoviesRemoteDataSourceImpl.kt:128]: seprate this and format extension function , every extension  in line to more readable

File: repository/src/main/java/com/cairosquad/repository/movie/MovieRepositoryImpl.kt [hunk 19]
```diff
@@ -6,155 +6,206 @@ import com.cairosquad.entity.Artist
...
+    }
+
+    override suspend fun getMoviesByQuery(query: String, page: Int): List<Movie> {
+        return getMovies(
+            remoteFetcher = { moviesRemoteDataSource.getMoviesByQuery(query, page) },
+            requestCache = getRequestOfSearchedMovies(query, page)
+        )
+    }
+
+    private suspend fun getMovies(
```
Comment (ElsayedMagdy122) [inline repository/src/main/java/com/cairosquad/repository/movie/MovieRepositoryImpl.kt:149]: seprate , it is not readable

File: repository/src/main/java/com/cairosquad/repository/movie/MovieRepositoryImpl.kt [hunk 20]
```diff
@@ -6,155 +6,206 @@ import com.cairosquad.entity.Artist
...
-    }
-
-    override suspend fun getNowPlayingMovies(page: Int,categoryId: String?): List<Movie> {
-        return tryToCall {
-            remoteMovieDataSource.getNowPlayingMovies(page,categoryId).map { it.toEntity() }
-        }
+            }
+    }
+
+    override suspend fun getMovieById(movieId: Long): Movie {
```
Comment (ElsayedMagdy122) [inline repository/src/main/java/com/cairosquad/repository/movie/MovieRepositoryImpl.kt]: same seprate and make paramete id only it is simple

File: repository/src/main/java/com/cairosquad/repository/movie/data_source/local/CacheDtoMappers.kt [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+    return map { it.toEntity() }
+}
+
+fun Movie.toCacheDto(): MovieCacheDto {
+    return MovieCacheDto(
+        movieWithoutGenre = MovieWithoutGenreCacheDto(
+            id = id.toLong(),
+            title = title,
+            posterPath = posterPath,
+            voteAverage = rating,
```
Comment (ElsayedMagdy122) [inline repository/src/main/java/com/cairosquad/repository/movie/data_source/local/CacheDtoMappers.kt]: why voteAverage ? why the naming not rating

File: repository/src/main/java/com/cairosquad/repository/movie/data_source/local/CacheDtoMappers.kt [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+
+@JvmName("toEntityMovie")
+fun List<MovieCacheDto>.toEntityList(): List<Movie> {
+    return map { it.toEntity() }
+}
+
+fun Movie.toCacheDto(): MovieCacheDto {
+    return MovieCacheDto(
+        movieWithoutGenre = MovieWithoutGenreCacheDto(
+            id = id.toLong(),
```
Comment (ElsayedMagdy122) [inline repository/src/main/java/com/cairosquad/repository/movie/data_source/local/CacheDtoMappers.kt]: why to long , by default it is long ?

File: repository/src/main/java/com/cairosquad/repository/movie/data_source/local/dto/MovieWithoutGenreCacheDto.kt [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+    @ColumnInfo(name = "poster_path")
+    val posterPath: String?,
+    @ColumnInfo(name = "release_date")
+    val releaseDate: Long,
+    @ColumnInfo(name = "runtime")
+    val runtime: Int?,
+    @ColumnInfo(name = "trailerPath")
+    val trailerPath: String,
+    @ColumnInfo(name = "vote_average")
+    val voteAverage: Float,
```
Comment (ElsayedMagdy122) [inline repository/src/main/java/com/cairosquad/repository/movie/data_source/local/dto/MovieWithoutGenreCacheDto.kt]: make naming : rating

File: repository/src/main/java/com/cairosquad/repository/series/SeriesRepositoryImpl.kt [hunk 24]
```diff
@@ -9,103 +9,205 @@ import com.cairosquad.entity.Review
...
-
-    override suspend fun getFreeToWatchSeries(page: Int,categoryId : String?): List<Series> {
-        return tryToCall {
-            remoteSeriesDataSource.getFreeToWatchSeries(page,categoryId).map { it.toEntity() }
-        }
+    private suspend fun getSeries(
+        remoteFetcher: suspend () -> List<SeriesRemoteDto>,
+        requestCache: String
+    ):List<Series> {
+        seriesLocalDataSource.deleteExpiredCache(Date().time - CACHE_EXPIRATION_MILLIS)
```
Comment (ElsayedMagdy122) [inline repository/src/main/java/com/cairosquad/repository/series/SeriesRepositoryImpl.kt:141]: you can make private fuction helper to decrase duplicated code and seprate it

File: repository/src/main/java/com/cairosquad/repository/series/SeriesRepositoryImpl.kt [hunk 25]
```diff
@@ -9,103 +9,205 @@ import com.cairosquad.entity.Review
...
+        return seriesLocalDataSource
+            .getSeriesByRequest(request = requestCache)
```
Comment (ElsayedMagdy122) [inline repository/src/main/java/com/cairosquad/repository/series/SeriesRepositoryImpl.kt]: same

File: repository/src/main/java/com/cairosquad/repository/series/SeriesRepositoryImpl.kt [hunk 26]
```diff
@@ -9,103 +9,205 @@ import com.cairosquad.entity.Review
...
+                        seriesLocalDataSource.insertRequestWithSeries(
+                            series.toRequestWithSeriesCacheDto(
+                                request = requestCache
+                            )
+                        )
+                    }
+            }
+    }
+
+    override suspend fun getSeriesById(seriesId: Long): Series {
```
Comment (ElsayedMagdy122) [inline repository/src/main/java/com/cairosquad/repository/series/SeriesRepositoryImpl.kt]: same

File: repository/src/main/java/com/cairosquad/repository/series/data_source/local/CacheDtoMappers.kt [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+    return RequestWithSeriesCacheDto(
+        request = RequestCacheDto(request = request),
+        series = this.map { it.toCacheDto() }
+    )
+}
+
+
+fun SeriesCacheDto.toEntity(): Series {
+    return Series(
+        id = seriesWithoutGenre.id.toLong(),
```
Comment (ElsayedMagdy122) [inline repository/src/main/java/com/cairosquad/repository/series/data_source/local/CacheDtoMappers.kt]: it is by default long

File: repository/src/main/java/com/cairosquad/repository/utils/sharedDto/local/RequestWithReviewsCacheDto.kt [hunk 28]
```diff
@@ -0,0 +1,16 @@
+package com.cairosquad.repository.utils.sharedDto.local
+
+import androidx.room.Embedded
+import androidx.room.Junction
+import androidx.room.Relation
+
+data class RequestWithReviewsCacheDto(
+    @Embedded
```
Comment (ElsayedMagdy122) [inline repository/src/main/java/com/cairosquad/repository/utils/sharedDto/local/RequestWithReviewsCacheDto.kt]: can you explain to my why Embded

File: repository/src/test/java/com/cairosquad/repository/series/SeriesRepositoryImplTest.kt [hunk 29]
```diff
@@ -14,23 +14,23 @@ import org.junit.Test
...
         seriesRepository = SeriesRepositoryImpl(
-            remoteSeriesDataSource = remoteSeriesDataSource
+            seriesRemoteDataSource = seriesRemoteDataSource
         )
     }
 
     @Test
     fun `should return fake series when getSeries is called`() = runTest {
-        coEvery { remoteSeriesDataSource.getSeries(1399L) } returns SeriesDetailsRemoteDto()
+        coEvery { seriesRemoteDataSource.getSeriesById(1399L) } returns SeriesDetailsRemoteDto()
```
Comment (ElsayedMagdy122) [inline repository/src/test/java/com/cairosquad/repository/series/SeriesRepositoryImplTest.kt]: herd coded , create variable and pass it

**Comment by ElsayedMagdy122** [comment]
finally , fix test errors

File: domain/src/main/java/com/cairosquad/domain/repository/SearchRepository.kt (hunk 1 above)
Comment (Al-Qassim) [reply domain/src/main/java/com/cairosquad/domain/repository/SearchRepository.kt]: moved it

File: domain/src/main/java/com/cairosquad/domain/repository/SeriesRepository.kt (hunk 2 above)
Comment (Al-Qassim) [reply domain/src/main/java/com/cairosquad/domain/repository/SeriesRepository.kt]: done

File: domain/src/main/java/com/cairosquad/domain/repository/SeriesRepository.kt (hunk 3 above)
Comment (Al-Qassim) [reply domain/src/main/java/com/cairosquad/domain/repository/SeriesRepository.kt]: done

File: domain/src/main/java/com/cairosquad/domain/usecase/movies/GetMovieDetailsUseCase.kt (hunk 4 above)
Comment (Al-Qassim) [reply domain/src/main/java/com/cairosquad/domain/usecase/movies/GetMovieDetailsUseCase.kt]: done

File: domain/src/main/java/com/cairosquad/domain/usecase/search/SearchUseCase.kt (hunk 5 above)
Comment (Al-Qassim) [reply domain/src/main/java/com/cairosquad/domain/usecase/search/SearchUseCase.kt]: done

File: domain/src/main/java/com/cairosquad/domain/usecase/search/SearchUseCase.kt (hunk 6 above)
Comment (Al-Qassim) [reply domain/src/main/java/com/cairosquad/domain/usecase/search/SearchUseCase.kt]: done

File: domain/src/main/java/com/cairosquad/domain/usecase/series/GetSeriesDetailsUseCase.kt (hunk 7 above)
Comment (Al-Qassim) [reply domain/src/main/java/com/cairosquad/domain/usecase/series/GetSeriesDetailsUseCase.kt]: done

File: domain/src/test/java/com/cairosquad/domain/usecase/movies/GetMoviesDetailsUseCaseTest.kt (hunk 8 above)
Comment (Al-Qassim) [reply domain/src/test/java/com/cairosquad/domain/usecase/movies/GetMoviesDetailsUseCaseTest.kt]: I didn't fix the tests yet

File: local/src/main/java/com/cairosquad/local/cache/movie/MoviesCacheDao.kt (hunk 9 above)
Comment (Al-Qassim) [reply local/src/main/java/com/cairosquad/local/cache/movie/MoviesCacheDao.kt]: done

File: local/src/main/java/com/cairosquad/local/cache/movie/MoviesCacheDao.kt (hunk 10 above)
Comment (Al-Qassim) [reply local/src/main/java/com/cairosquad/local/cache/movie/MoviesCacheDao.kt]: done

File: local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt (hunk 11 above)
Comment (Al-Qassim) [reply local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt]: done

File: local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt (hunk 12 above)
Comment (Al-Qassim) [reply local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt]: done

File: local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt (hunk 13 above)
Comment (Al-Qassim) [reply local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt]: done

File: local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt (hunk 14 above)
Comment (Al-Qassim) [reply local/src/main/java/com/cairosquad/local/cache/movie/MoviesLocalDataSourceImpl.kt:79]: no need for that actually, it's just one function calling 7 other function

File: local/src/main/java/com/cairosquad/local/cache/series/SeriesCacheDao.kt (hunk 15 above)
Comment (Al-Qassim) [reply local/src/main/java/com/cairosquad/local/cache/series/SeriesCacheDao.kt]: renaming done

File: local/src/main/java/com/cairosquad/local/cache/series/SeriesLocalDataSourceImpl.kt (hunk 16 above)
Comment (Al-Qassim) [reply local/src/main/java/com/cairosquad/local/cache/series/SeriesLocalDataSourceImpl.kt]: it won't make it batter honestly

File: local/src/main/java/com/cairosquad/local/cache/series/SeriesLocalDataSourceImpl.kt (hunk 17 above)
Comment (Al-Qassim) [reply local/src/main/java/com/cairosquad/local/cache/series/SeriesLocalDataSourceImpl.kt:76]: same

File: remote/src/main/java/com/cairosquad/remote/movie/MoviesRemoteDataSourceImpl.kt (hunk 18 above)
Comment (Al-Qassim) [reply remote/src/main/java/com/cairosquad/remote/movie/MoviesRemoteDataSourceImpl.kt:128]: let it as it is, so the test coverage will be higher, and it's actually not that bad for readability

File: repository/src/main/java/com/cairosquad/repository/movie/MovieRepositoryImpl.kt (hunk 19 above)
Comment (Al-Qassim) [reply repository/src/main/java/com/cairosquad/repository/movie/MovieRepositoryImpl.kt:149]: can you help me refactor it?

File: repository/src/main/java/com/cairosquad/repository/movie/MovieRepositoryImpl.kt (hunk 20 above)
Comment (Al-Qassim) [reply repository/src/main/java/com/cairosquad/repository/movie/MovieRepositoryImpl.kt]: done

File: repository/src/main/java/com/cairosquad/repository/movie/data_source/local/CacheDtoMappers.kt (hunk 21 above)
Comment (Al-Qassim) [reply repository/src/main/java/com/cairosquad/repository/movie/data_source/local/CacheDtoMappers.kt]: done

File: repository/src/main/java/com/cairosquad/repository/movie/data_source/local/CacheDtoMappers.kt (hunk 22 above)
Comment (Al-Qassim) [reply repository/src/main/java/com/cairosquad/repository/movie/data_source/local/CacheDtoMappers.kt]: removed to long, done

File: repository/src/main/java/com/cairosquad/repository/series/SeriesRepositoryImpl.kt (hunk 24 above)
Comment (Al-Qassim) [reply repository/src/main/java/com/cairosquad/repository/series/SeriesRepositoryImpl.kt:141]: done

File: repository/src/main/java/com/cairosquad/repository/series/SeriesRepositoryImpl.kt (hunk 25 above)
Comment (Al-Qassim) [reply repository/src/main/java/com/cairosquad/repository/series/SeriesRepositoryImpl.kt]: same

File: repository/src/main/java/com/cairosquad/repository/series/SeriesRepositoryImpl.kt (hunk 26 above)
Comment (Al-Qassim) [reply repository/src/main/java/com/cairosquad/repository/series/SeriesRepositoryImpl.kt]: done

File: repository/src/main/java/com/cairosquad/repository/series/data_source/local/CacheDtoMappers.kt (hunk 27 above)
Comment (Al-Qassim) [reply repository/src/main/java/com/cairosquad/repository/series/data_source/local/CacheDtoMappers.kt]: done

File: repository/src/main/java/com/cairosquad/repository/utils/sharedDto/local/RequestWithReviewsCacheDto.kt (hunk 28 above)
Comment (Al-Qassim) [reply repository/src/main/java/com/cairosquad/repository/utils/sharedDto/local/RequestWithReviewsCacheDto.kt]: this is how to make a many to many relationship in room

File: repository/src/test/java/com/cairosquad/repository/series/SeriesRepositoryImplTest.kt (hunk 29 above)
Comment (Al-Qassim) [reply repository/src/test/java/com/cairosquad/repository/series/SeriesRepositoryImplTest.kt]: didn't fix the tests yet

File: repository/src/main/java/com/cairosquad/repository/utils/sharedDto/local/RequestWithReviewsCacheDto.kt (hunk 28 above)
Comment (ElsayedMagdy122) [reply repository/src/main/java/com/cairosquad/repository/utils/sharedDto/local/RequestWithReviewsCacheDto.kt]: ok , you can create resolve for it

File: repository/src/main/java/com/cairosquad/repository/movie/MovieRepositoryImpl.kt (hunk 19 above)
Comment (ElsayedMagdy122) [reply repository/src/main/java/com/cairosquad/repository/movie/MovieRepositoryImpl.kt:149]: @Al-Qassim 
Ex :-
```
private suspend fun getCachedMovies(request: String): List<Movie> {
    return moviesLocalDataSource
        .getMoviesByRequest(request)
        .toEntityList()
}

private suspend fun fetchAndCacheMovies(
    remoteFetcher: suspend () -> List<MovieRemoteDto>,
    requestCache: String
): List<Movie> = tryToCall {
    val genres = fetchGenres()
    val movies = remoteFetcher().map { it.toEntity(genres) }

    cacheMovies(requestCache, movies)

    movies
}

private suspend fun fetchGenres(): List<Genre> {
    return moviesRemoteDataSource
        .getMoviesGenres()
        .map { it.toEntity() }
}

private suspend fun cacheMovies(
    requestCache: String,
    movies: List<Movie>
) {
    val cacheDto = movies.toRequestWithMoviesCacheDto(request = requestCache)
    moviesLocalDataSource.insertRequestWithMovies(cacheDto)
}
```

File: repository/src/main/java/com/cairosquad/repository/movie/data_source/local/dto/MovieWithoutGenreCacheDto.kt (hunk 23 above)
Comment (Al-Qassim) [reply repository/src/main/java/com/cairosquad/repository/movie/data_source/local/dto/MovieWithoutGenreCacheDto.kt]: done

File: app/src/main/java/com/cairosquad/movio/di/localDataSourceModule.kt [hunk 30]
```diff
@@ -27,31 +36,59 @@ val localDataSourceModule = module {
...
+    single<SeriesLocalDataSource> {
+        SeriesLocalDataSourceImpl(get(), get(), get(), get())
+    }
+
+    single<SeasonEpisodeLocalDataSource> {
+        SeasonEpisodeLocalDataSourceImpl(get(), get())
+    }
+
+    single<ArtistsLocalDataSource> {
+        ArtistsLocalDataSourceImpl(get(), get())
```
Comment (gehadepada) [inline app/src/main/java/com/cairosquad/movio/di/localDataSourceModule.kt:92]: you can use this :
 it single<MoviesLocalDataSource> ::MoviesLocalDataSourceImpl
single<SeriesLocalDataSource> ::SeriesLocalDataSourceImpl
as bareq sayed is not practical to use get

File: app/src/main/java/com/cairosquad/movio/di/repositoryModule.kt [hunk 31]
```diff
@@ -18,19 +18,18 @@ val repositoryModule = module {
...
 
     single<MoviesRepository> {
-        MovieRepositoryImpl(get(), get(), get())
+        MovieRepositoryImpl(get(), get())
     }
 
     single<SeriesRepository> {
-        SeriesRepositoryImpl(get())
+        SeriesRepositoryImpl(get(), get(), get())
     }
```
Comment (gehadepada) [inline app/src/main/java/com/cairosquad/movio/di/repositoryModule.kt:30]: same

File: domain/src/test/java/com/cairosquad/domain/usecase/ManageArtistUseCaseTest.kt [hunk 32]
```diff
@@ -1,63 +1,74 @@
...
 
         val result = useCase.getSeriesOfArtist(312L)
 
-        assertThat(result).containsExactly(series1, series2)
-        coVerify(exactly = 1) { artistsRepository.getSeriesOfArtist(312L) }
+        Truth.assertThat(result).containsExactly(series1, series2)
+        coVerify(exactly = 1) { seriesRepository.getSeriesOfArtist(312L) }
     }
 
     // 🔥 Edge Case
```
Comment (gehadepada) [inline domain/src/test/java/com/cairosquad/domain/usecase/ManageArtistUseCaseTest.kt]: delete this comment

File: local/src/main/java/com/cairosquad/local/cache/series/SeriesLocalDataSourceImpl.kt [hunk 33]
```diff
@@ -0,0 +1,88 @@
...
+        seriesCacheDao.deleteExpiredSeriesWithoutGenreCache(timestamp)
+        seriesCacheDao.deleteCrossRefForNonExistingCacheCodeAndSeriesCache()
+
+        genreDao.deleteExpiredGenreOfSeriesCache(timestamp)
+        seriesCacheDao.deleteCrossRefForNonExistingSeriesAndGenreCache()
+
+        reviewDao.deleteExpiredReviewCache(timestamp)
+        reviewDao.deleteCrossRefForNonexistingRequestReviewCache()
+    }
+}
```
Comment (gehadepada) [inline local/src/main/java/com/cairosquad/local/cache/series/SeriesLocalDataSourceImpl.kt:88]: yoy can separate this function into small function

Comment (Al-Qassim) [reply local/src/main/java/com/cairosquad/local/cache/series/SeriesLocalDataSourceImpl.kt:88]: no, it's just 1 function calling 7 other functions

File: repository/src/test/java/com/cairosquad/repository/utils/mappers/MovieCacheDtoExtensionsTest.kt
Comment (iZyadNasser) [inline repository/src/test/java/com/cairosquad/repository/utils/mappers/MovieCacheDtoExtensionsTest.kt:1]: remove commented file

File: app/src/main/java/com/cairosquad/movio/di/localDataSourceModule.kt (hunk 30 above)
Comment (Al-Qassim) [reply app/src/main/java/com/cairosquad/movio/di/localDataSourceModule.kt:92]: I can't

File: app/src/main/java/com/cairosquad/movio/di/repositoryModule.kt (hunk 31 above)
Comment (Al-Qassim) [reply app/src/main/java/com/cairosquad/movio/di/repositoryModule.kt:30]: same

File: domain/src/test/java/com/cairosquad/domain/usecase/ManageArtistUseCaseTest.kt (hunk 32 above)
Comment (Al-Qassim) [reply domain/src/test/java/com/cairosquad/domain/usecase/ManageArtistUseCaseTest.kt]: ok

**Review by iZyadNasser — APPROVED** [review]

**Comment by MostafaMohamed2002** [comment]
great work

**Review by MostafaMohamed2002 — APPROVED** [review]

**Review by gehadepada — APPROVED** [review]
